File: util/anim/AnimCI.java
package common.util.anim;

import common.CommonStatic;
import common.io.json.JsonClass;
import common.pack.Source;
import common.pack.Source.ResourceLocation;
import common.system.VImg;
import common.system.fake.FakeImage;
import common.system.fake.FakeImage.Marker;

import java.util.List;

@JsonClass.JCGeneric(ResourceLocation.class)
public class AnimCI extends AnimU<AnimCI.AnimCIKeeper> {

	protected static class AnimCIKeeper implements AnimU.ImageKeeper {

		public final Source.AnimLoader loader;
		private FakeImage num;
		private boolean ediLoaded = false;
		private VImg edi;
		private VImg uni;

		private AnimCIKeeper(Source.AnimLoader al) {
			loader = al;
		}

		@Override
		public VImg getEdi() {
			if (ediLoaded)
				return edi;
			ediLoaded = true;
			edi = loader.getEdi();
			if (edi != null)
				edi.mark(Marker.EDI);
			return edi;
		}

		@Override
		public ImgCut getIC() {
			return loader.getIC();
		}

		@Override
		public MaAnim[] getMA() {
			return loader.getMA();
		}

		@Override
		public MaModel getMM() {
			return loader.getMM();
		}

		public ResourceLocation getName() {
			return loader.getName();
		}

		@Override
		public FakeImage getNum() {
			if (num != null && num.bimg() != null && num.isValid())
				return num;
			return num = loader.getNum();
		}

		public int getStatus() {
			return loader.getStatus();
		}

		@Override
		public VImg getUni() {
			if (uni != null)
				return uni;
			uni = loader.getUni();
			if (uni != null)
				uni.mark(Marker.UNI);
			else
				uni = CommonStatic.getBCAssets().slot[0];
			return uni;
		}

		public void setEdi(VImg vedi) {
			edi = vedi;

			if (vedi != null)
				vedi.mark(Marker.EDI);

			ediLoaded = true;
		}

		public void setNum(FakeImage fimg) {
			num = fimg;
		}

		public void setUni(VImg vuni) {
			uni = vuni;
			uni.mark(Marker.UNI);
		}

		@Override
		public void unload() {
			if(num != null) {
				num.unload();

				num = null;
			}
		}

		@Override
		public boolean validate(AnimationType type) {
			return loader.validate(type);
		}

		@Override
		public List<String> collectInvalidAnimation(AnimationType type) {
			return loader.collectInvalidAnimation(type);
		}
	}

	@JsonClass.JCIdentifier
	public ResourceLocation id;

	public AnimCI(Source.AnimLoader acl) {
		super(new AnimCIKeeper(acl));
		id = loader.getName();
	}

	@Override
	public void load() {
		try {
			super.load();
			if (getEdi() != null)
				getEdi().check();
			if (getUni() != null)
				getUni().check();
		} catch (Exception e) {
			e.printStackTrace();
			CommonStatic.def.save(false, true);
		}
		validate();
	}

	@Override
	public boolean cantLoadAll(ImageKeeper.AnimationType type) {
		return !loader.validate(type);
	}

	@Override
	public List<String> collectInvalidAnimation(ImageKeeper.AnimationType type) {
		return loader.collectInvalidAnimation(type);
	}

	@Override
	public String toString() {
		return id.id;
	}

}

File: util/anim/AnimCI.kt
package common.util.anim

import common.CommonStatic
import common.pack.Source.AnimLoader
import common.pack.Source.ResourceLocation
import common.system.VImg
import common.system.fake.FakeImage
import common.system.fake.FakeImage.Marker
import common.util.anim.AnimCI.AnimCIKeeper

open class AnimCI(acl: AnimLoader) : AnimU<AnimCIKeeper?>(AnimCIKeeper(acl)) {
    protected class AnimCIKeeper(al: AnimLoader) : ImageKeeper {
        val loader: AnimLoader
        private var num: FakeImage? = null
        private var ediLoaded = false
        private var edi: VImg? = null
        private var uni: VImg? = null
        override fun getEdi(): VImg? {
            if (ediLoaded) return edi
            ediLoaded = true
            edi = loader.getEdi()
            if (edi != null) edi.mark(Marker.EDI)
            return edi
        }

        override fun getIC(): ImgCut {
            return loader.getIC()
        }

        override fun getMA(): Array<MaAnim?> {
            return loader.getMA()
        }

        override fun getMM(): MaModel {
            return loader.getMM()
        }

        fun getName(): ResourceLocation {
            return loader.getName()
        }

        override fun getNum(): FakeImage? {
            return if (num != null && num.bimg() != null && num.isValid()) num else loader.getNum().also { num = it }
        }

        fun getStatus(): Int {
            return loader.getStatus()
        }

        override fun getUni(): VImg? {
            if (uni != null) return uni
            uni = loader.getUni()
            if (uni != null) uni.mark(Marker.UNI) else uni = CommonStatic.getBCAssets().slot.get(0)
            return uni
        }

        fun setEdi(vedi: VImg?) {
            edi = vedi
            if (vedi != null) vedi.mark(Marker.EDI)
            ediLoaded = true
        }

        fun setNum(fimg: FakeImage?) {
            num = fimg
        }

        fun setUni(vuni: VImg?) {
            uni = vuni
            uni.mark(Marker.UNI)
        }

        override fun unload() {}

        init {
            loader = al
        }
    }

    var id: ResourceLocation
    override fun load() {
        try {
            super.load()
            if (getEdi() != null) getEdi().check()
            if (getUni() != null) getUni().check()
        } catch (e: Exception) {
            e.printStackTrace()
            CommonStatic.def.exit(false)
        }
        validate()
    }

    override fun toString(): String {
        return id.id
    }

    init {
        id = loader.getName()
    }
}
